packages/mask/src/plugins/CryptoartAI/hooks/usePurchaseCallback.ts

```diff
@@ -0,0 +1,89 @@
+import { useCallback } from 'react'
+import BigNumber from 'bignumber.js'
+import type { NonPayableTx } from '@masknet/web3-contracts/types/types'
+import {
+    TransactionEventType,
+    TransactionStateType,
+    useTransactionState,
+    useAccount,
+    useChainId,
+} from '@masknet/web3-shared-evm'
+import { useCryptoArtAI_Contract } from './useCryptoArtAI_Contract'
+
+export function usePurchaseCallback(editionNumber: number, priceInWei: number) {
+    const account = useAccount()
+    const chainId = useChainId()
+    const { knownOriginDigitalAssetV2_contract } = useCryptoArtAI_Contract()
+    const [purchaseState, setPurchaseState] = useTransactionState()
+
+    const purchaseCallback = useCallback(async () => {
+        if (!knownOriginDigitalAssetV2_contract) {
+            setPurchaseState({
+                type: TransactionStateType.UNKNOWN,
+            })
+            return
+        }
+
+        // start waiting for provider to confirm tx
+        setPurchaseState({
+            type: TransactionStateType.WAIT_FOR_CONFIRMING,
+        })
+
+        // estimate gas and compose transaction
+        const config = {
+            from: account,
+            value: new BigNumber(priceInWei).toFixed(),
+            gas: await knownOriginDigitalAssetV2_contract.methods
+                .purchase(editionNumber)
+                .estimateGas({
+                    from: account,
+                    value: new BigNumber(priceInWei).toFixed(),
+                })
+                .catch((error) => {
+                    setPurchaseState({
+                        type: TransactionStateType.FAILED,
+                        error,
+                    })
+                    throw error
+                }),
+        }
+
+        // send transaction and wait for hash
+        return new Promise<void>(async (resolve, reject) => {
+            knownOriginDigitalAssetV2_contract.methods
+                .purchase(editionNumber)
+                .send(config as NonPayableTx)
+                .on(TransactionEventType.RECEIPT, (receipt) => {
```

Comment:
It also needs to listen to the `TransactionEventType.HASH` event. Cause the rest events may not emit totally.